# Python Assignment 3

⏱⏱⏱**Due: 11/02/2023 11:59 PM** ⏱⏱⏱

Please submit your notebook files and upload them to your Git repository. Don't forget to include a link to your Git repository when submitting this file on Brightspace. **Please include all outputs in your notebook, including plots, printouts, etc..**

Collaboration is permitted and encouraged; however, it's essential to independently produce and submit your own work. If you collaborate with others, please ensure that you provide their names in the designated section.

Collaborators:_____

##**For Questions 2a and 2b, you can choose one of them to complete, if you completed both, only Questions 2a will be graded.**

**This assignment is out of 150 points; surplus points can be applied to other assignments (including Matlab and C/C++). However, it cannot be applied to your final project grade, and your assignment grade will not excced 75% of your total grade.**

## 1. Good presentation of your code and outputs; submitting your files to Github (10 pts)

Ensure thorough and clear comments within your code to explain its functionality.

Enhance your plots by adding titles, legends, and axis labels where appropriate to provide clarity.

If necessary, employ LaTex notation to present mathematical explanations within the notebook.

Divide your code into multiple blocks or cells in Jupyter Notebook and provide descriptive explanations for each block to improve readability.

As part of your submission, **include the notebook files and upload them to your Git repository. Additionally, remember to provide a link to your Git repository when submitting the files on Brightspace.** Do not submit a compressed file (.rar, .zip, etc..), submit files separately.

If you are tasked with implementing a numerical algorithm, it is expected that you would not rely on pre-existing methods. For example, if you were asked to solve $Ax=b$, you should write your own program to solve, not just by calling numpy.linalg.solve.



## 2a. Numerical Integration (30 pts)

For a domain discretized into $N$ equally spaced panels:
$$
\Delta x=\frac{b-a}{N}.
$$
The approximation to the integral is
$$\begin{aligned}
\int_a^b f(x) d x & \approx \frac{\Delta x}{2} \sum_{k=1}^N\left(f\left(x_{k-1}\right)+f\left(x_k\right)\right) \\
& =\frac{\Delta x}{2}\left(f\left(x_0\right)+2 f\left(x_1\right)+2 f\left(x_2\right)+2 f\left(x_3\right)+\cdots+2 f\left(x_{N-1}\right)+f\left(x_N\right)\right) \\
& =\Delta x\left(\sum_{k=1}^{N-1} f\left(x_k\right)+\frac{f\left(x_N\right)+f\left(x_0\right)}{2}\right) .
\end{aligned}$$
This is the trapezoid rule scheme.


Given the function $f(x) = 3x^2 + 2x +2$, your tasks are:

1. Carry out the integration symbolically using the SymPy library. Output the result (i.e. print out the result).
2. Implement the trapezoid rule to estimate the integral $\int_{-4}^6 f(x) dx$ using the following values of $N = 10, 20, 40, 80, 160, 320, 640, 1280$. Output the results (i.e. print out the results).
3. Plot the absolute error (absolute difference between the numerical approximation and the SymPy result) against N.





In [ ]:
#import relevant Python libraries
import sympy as sym
import numpy as np

#define x as a symbol for symbolic computation
x = sym.Symbol('x')
print(f'Symoblic integration result: {sym.integrate(3*x**2 + 2*x + 2)}')

#define a function to perform numerical integration via the trapezoidal rule
def trapezoidal_rule(f, a, b, N):
    #f is a function
    #a and b are bounds
    #N is the number of trapezoids

    #define x-axis distance between left-sides of trapezoids
    dx = (b - a) / N
    #define x-axis points that denote where trapezoids' sides connect w/ x-axis
    xs = np.linspace(a, b, N)
    #return result of numerical integration
    return dx * (np.sum(f(xs[:N-1]) + (f(xs[-1]) + f(xs[0]))/2))

#define function that will be integrated
f = lambda x : 3*x**2 + 2*x + 2
#define min and max values of the domain of integration
a, b = -4, 6
#perform numerical integration with increasing values of N
for N in (10, 20, 40, 80, 160, 320, 640, 1280):
    print(f'Trapezoidal rule-integration with N={N}: {trapezoidal_rule(f, a, b, N)}')

Symoblic integration result: x**3 + x**2 + 2*x
Trapezoidal rule-integration with N=10: 991.5555555555557
Trapezoidal rule-integration with N=20: 1064.3157894736842
Trapezoidal rule-integration with N=40: 1101.8205128205127
Trapezoidal rule-integration with N=80: 1120.8291139240507
Trapezoidal rule-integration with N=160: 1130.3946540880504
Trapezoidal rule-integration with N=320: 1135.1923981191221
Trapezoidal rule-integration with N=640: 1137.5949726134586
Trapezoidal rule-integration with N=1280: 1138.7971804143863


## 2b. Numerical Differentiation (30 pts)

Your tasks:
1. Find the closed form expression of $f_{xy}(x,y)$, i.e. $\frac{d^2}{dxdy}f(x,y)$, for $f(x,y) = \left(sin^2\left(x\right)e^xcos(y)\right)$ using `sympy`. Output the result.
2. Find $f_{xy}(2,3)$ with `sympy` and take 15 significant digits, use it as the "groud truth solution". Output the result.
3. Approximate $f_{xy}(2,3)$ using central difference approximation for the following values of $h = 0.1, 0.01, 0.001, 0.0001$. Output the results.
4. Plot the absolute error (absolute difference between the numerical approximation and the SymPy result) against the step size $h$. `plt.gca().invert_xaxis() ` might help.

Recall from the lecture:
$$f_{x y}(x, y) \approx \frac{f(x+h, y+k)-f(x+h, y-k)-f(x-h, y+k)+f(x-h, y-k)}{4 h k}.$$
For simplicity, we set $k = h$.

## 3. Pandas I (15 pts)

1. Create a 3x4 (3 rows by 4 columns) pandas DataFrame with the columns named after the following Long Island towns: Hempstead, Babylon, Islip, and Brookhaven. The rows represent 'Population in 2099', 'Population in 2300', 'Population in 2400'. Fill each of the 12 cells in the DataFrame with a random integer from 1000 to 10000, inclusive. `np.random.randint()` might be helpful.
For Example:

```
                    Hempstead  Babylon  Islip  Brookhaven
Population in 2099       2931     8043   8414        8661
Population in 2300       5444     9227   7393        8007
Population in 2400       1660     7977   4730        2940
```



2. Output the following:
  - The entire DataFrame.
  - The value in the cell of row #1 (indexing starts with 0) under the Hempstead column.

3. Add a new column named Riverhead. Populate this column with the sum of the respective row values from the Islip and Brookhaven columns. Output the entire DataFrame again.

In [ ]:
import pandas as pd

#create dataset
data = {
    'Hempstead' : (2931, 5444, 1660),
    'Babylon'   : (8043, 9227, 7977),
    'Islip'     : (8414, 7393, 4730),
    'Brookhaven': (8661, 8007, 2940)}

#use dataset to create DataFrame
df = pd.DataFrame(
    data,
    index=pd.Index(('Population in 2099', 'Population in 2300', 'Population in 2400')),
    columns=pd.Index(('Hempstead', 'Babylon', 'Islip', 'Brookhaven')))

#print out the entire dataframe
print('DataFrame:\n',df,'\n')
#print out the requested value
print('Row #1 under Hempstead column:\n',df['Hempstead'][1],'\n')

#set a new column, 'Riverhead', to be defined as the sum of the Islip and Brookhaven columns
df['Riverhead'] = df['Islip'] + df['Brookhaven']

#print out the new DataFrame
print('DataFrame with Riverhead:\n',df,'\n')



DataFrame:
                     Hempstead  Babylon  Islip  Brookhaven
Population in 2099       2931     8043   8414        8661
Population in 2300       5444     9227   7393        8007
Population in 2400       1660     7977   4730        2940 

Row #1 under Hempstead column:
 5444 

DataFrame with Riverhead:
                     Hempstead  Babylon  Islip  Brookhaven  Riverhead
Population in 2099       2931     8043   8414        8661      17075
Population in 2300       5444     9227   7393        8007      15400
Population in 2400       1660     7977   4730        2940       7670 



## 4. Pandas II (15 pts)

0. Download the cvs file that comes with this assignment, and read it into a dataframe. You can also download this [csv file](https://media.githubusercontent.com/media/datablist/sample-csv-files/main/files/customers/customers-100.csv) with this link.
1. Arrange the data in alphabetical order based on the last name. Display the first few rows.
2. Count the number of customers whose subscription date is in 2021 (2021-01-01 to 2021-12-31). Report this number (e.g. print).

In [30]:
#connect Google Drive to Google Colab
from google.colab import drive
drive.mount('/content/drive/')

#read the CSV dataset from Google Drive into a DataFrame
df = pd.read_csv('/content/drive/MyDrive/ams595/python3/customers-100.csv')
#redefine the DataFrame as a version that is sorted by the 'Last Name' feature
df = df.sort_values('Last Name')
#print out the first 5 rows
df.head()

#print out the requested number of customers
print('Number of customers whose subscription date is in 2021:',df['Subscription Date'].str.contains('2021').value_counts()[True])


Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).
Number of customers whose subscription date is in 2021: 43


## 5. LU decomposition (50 + 30 pts)


Part A (50 points)
1. Write a function `LUdecomposition()` that takes as input a matrix $A\in \mathbb{R}^{n \times n}$. Perform Gaussian Elimination to have this function return two matrices $L$ and $U$ such that $A=LU$.
2. Your function should raise any relevant errors, for example, raise an error if a pure LU decomposition does not exist for $A$ (meaning that you have to perform pivoting, i.e., permutation of rows).
3. Create several random matrices for the purpose of testing your function. Obtain the $L$ and $U$ matrices, and for each random matrix obtain its L and U matrices and perform a matrix multiplication (`np.matmul()`) to determine whether the outcome closely resembles the original matrix $A$, allowing for slight differences attributable to precision errors.


Note: LU decomposition with Gaussian elimination should be covered in your undergraduate linear algebra class. It is also covered in AMS 510. If you do not know how, this 10 minute [video](https://www.youtube.com/watch?v=UlWcofkUDDU&ab_channel=Mathispower4u) explains it very well.

Part B (30 bonus points)
4. Not all square matrices have a pure LU decomposition. However, PLU decomposition always exists. Write another function `pivoting()` that takes as input a matrix $A$ and return the permutation/pivoting matrix $P$ such that $PA$ always has a pure LU decomposition.
5. Find an example $A$ for which `LUdecomposition()` raises an error. Then, use `pivoting()` to find $P$, and find the pure LU decomposition of $PA$. Verify that indeed $PA = LU$ with matrix multiplication (`np.matmul()`).


Hint: Please do not print out a giant matrix and visually examine the outcomes. Think of a way to verify your results and output a boolean value.


In [ ]:
import numpy as np

def LUdecomposition(A):
    #extract number of rows and columns in input matrix, A
    m, n = A.shape
    #raise error if input matrix is not square (nxn)
    if m != n:
        raise ValueError('Input matrix, A, is not square.')
    #initialize list which will store L matrices used to transform A into U
    Ls = []
    #iterate over columns (column major algorithm)
    for c in range(n):
        #initialize this iterations L matrix as the identity matrix
        L = np.identity(n)
        #iterate over rows below the "columnth" row
        for r in range(c+1, n):
            #raise error if dividing by zero due to a zero diagonal element in A
            if A[c, c] == 0:
                raise ValueError('Diagonal element in A is 0, which causes invalid division by 0.')
            #store the scaling factor used to induce a 0 into A to create U
            L[r,c] = -A[r,c] / A[c,c]
        #update with L to obtain the more-upper-triangular version of U
        A = np.einsum('ij,jk->ik', L, A)
        #store the L matrix used to convert A towards U
        Ls.append(L)

    #compute L^-1 by taking multiplying the individual L matrices together
    L_inv = Ls[0]
    for L in Ls[1:]:
        L_inv = np.einsum('ij,jk->ik', L, L_inv)

    #obtain the desired L matrix by inverting L^-1
    L = np.linalg.inv(L_inv)
    #obtain the desired U matrix by extracting the A matrix obtained by subsequent L matrices
    U = A
    #return L and U
    return L, U

# A = np.array(((3, 1),(-7, 2)))
# L, U = LUdecomposition(A)
# print(L)
# print(U)
# print(np.matmul(L,U))

#let the random matrices tested by 4x4
n = 4
#test out the LU algorithm 3 times
for i in range(3):
    #create random nxn matrix
    A = np.random.random((n,n))
    #compute L and U
    L, U = LUdecomposition(A)
    #compute A from L and U
    A_LU = np.matmul(L, U)
    #print relative errors of each element in A
    print(np.absolute((A - A_LU) / A), '\n')


[[0.00000000e+00 5.62967140e-16 3.46723245e-16 0.00000000e+00]
 [0.00000000e+00 2.36761042e-16 1.45817349e-15 1.62555544e-16]
 [0.00000000e+00 1.14899762e-16 3.42442196e-16 1.14035986e-15]
 [3.10520796e-16 1.28616286e-16 1.45203687e-16 4.43628536e-16]] 

[[1.60254967e-16 1.86282770e-16 0.00000000e+00 2.14263890e-16]
 [0.00000000e+00 1.32850470e-16 1.46369407e-16 1.80785116e-16]
 [1.74354670e-15 3.75521887e-16 5.11355145e-15 3.80533286e-16]
 [2.37929777e-16 1.75760650e-16 7.21946990e-16 0.00000000e+00]] 

[[1.32214702e-16 3.28602224e-16 2.59130771e-16 2.73655744e-14]
 [4.21869915e-16 1.28433787e-16 1.31491955e-16 1.34065024e-16]
 [1.22951489e-15 2.69600105e-16 3.67935792e-16 8.24692241e-16]
 [2.33948709e-16 2.55030310e-15 1.23576396e-15 5.41180267e-16]] 



In [ ]:
def pivoting(A):
    ...

Fact: LU and PLU decompositions are not unique.

## 6. Markov Chain (30 pts)

**BACKGROUND:** Consider a Markov chain transition matrix $P$ on a set of $n$ states, where $P_{i j}$ corresponds to the probability $\left(0 \leq P_{i j} \leq 1\right)$ to go from state $i$ to the state $j$, and each row is normalized so that
$$
\sum_{j=1}^n P_{i j}=1 .
$$

Let $p$ be a size- $n$ vector composed of the probability distribution over the $n$ states, where
$$
\sum_{j=1}^n p_j=1 .
$$

The transition rule corresponding to the transition matrix $P$ is $\hat{p}=P^T p$, where $\hat{p}$ corresponds to a new probability distribution and
$$
\sum_{j=1}^n \hat{p}_j=1 .
$$
TASK: Write a program that works with 5 states `only use the numpy library`. In particular,
0. Fix the random seed to be the last digit of your id: `numpy.random.seed()`.

1. Construct a random $5 \times 5$ matrix $P$, and normalize each row so that
$$
\sum_{j=1}^5 P_{i j}=1 .
$$

2. Construct a random size-5 vector $p$ and normalize it so that
$$
\sum_{j=1}^5 p_j=1 .
$$
Apply the transition rule 50 times to obtain $p_{50}$.
3. Compute the eigenvector $v$ of $P^T$ corresponding to the eigenvalue 1 (and numerically, the eigenvalue closest to $1)$, and then scale the eigenvector so that
$$
\sum_{j=1}^5 v_j=1 .
$$
This scaled eigenvector is known as the stationary distribution.

4. Compute the component wise difference between $p_{50}$ and the stationary distribution. Do they match with each other within $1 \mathrm{e}-5$ ?

HINTS: You may find the following functions useful in your implementation:
- np.random.rand
- np.dot
- np.linalg.eig

In [ ]:
#set seed to be the last digit of my ID
np.random.seed(8)

#create random 5x5 matrix that is row-normlized via softmax
n = 5
P = np.random.random((n,n))
#create a softmax function which converts arrays into probability distributions
softmax = lambda arr : np.exp(arr) / np.sum(np.exp(arr))
for r in range(n):
    #apply softmax to each row of P
    P[r,:] = softmax(P[r,:])

#create random 5x1 vector that is normlized via softmax
p = np.random.random(n)
p = softmax(p)

#apply transition 50 times to obtain p_50
#   compute eigendecomposition and use the fact that f(A) = U f(D) U^-1 to find P^50
eigvals, U = np.linalg.eig(P)
P_50 = np.matmul(np.matmul(U, np.diag(eigvals)**50), np.linalg.inv(U))
p_50 = np.matmul(P_50, p)

#eig returns eigenvalues in decreasing order,
#   so for a Markov matrix the first eigenvector corresponds to lambda = 1
#   all eigenvalues have magnitudes less than or equal to 1
eigenvec = U[:, 0]
stationary_dist = softmax(eigenvec)

#print the absolute error in the difference between p_50 and the stationary distribution
print(p_50 - stationary_dist)


[-8.77932411e-05 -8.77932411e-05 -8.77932411e-05 -8.77932411e-05
 -8.77932411e-05]


As shown by the array printed above, the component wise difference between $p_{50}$ and the stationary distribution ALMOST match with each other within $1 \mathrm{e}-5$. The absolute error is $-8.78\mathrm{e}-05$ for each element of the 5-vector, which is on the order of magnitude of $1\mathrm{e}-5$.